modules/core/shared/src/main/scala/Trace.scala

```diff
@@ -205,15 +205,17 @@ object Trace {
         Kleisli.liftF(trace.traceUri)
     }
 
-  implicit def liftStateT[F[_]: Monad, S](implicit trace: Trace[F]): Trace[StateT[F, S, *]] =
+  implicit def liftStateT[F[_]: MonadCancelThrow, S](implicit trace: Trace[F]): Trace[StateT[F, S, *]] =
     new Trace[StateT[F, S, *]] {
-
       def put(fields: (String, TraceValue)*): StateT[F, S, Unit] =
         StateT.liftF(trace.put(fields: _*))
 
       def kernel: StateT[F, S, Kernel] =
         StateT.liftF(trace.kernel)
 
+      def spanR[A](name: String)(r: Resource[StateT[F, S, *], A]): Resource[StateT[F, S, *], A] =
+        trace.spanR(name)(Resource.unit).mapK(StateT.liftK[F, S]) *> r
```

Comment:
I haven't fully grasped the implications of these definitions, but is this much different than if we tried to implement `spanR(name: String): Resource[F, Unit]` instead?

Reply:
I haven't tested this, but ...

- If the transformed instance is "stateful" (parents `acquire`, `use`, and `release`), so are these, and would be nicer expressed as `spanR(name: String): Resource[F, Unit]`.

- If the transformed instance is "local" (parents `acquire` and `release` but not `use`), we still span the resource, but parent none of the three phases.

- Local and stateful tracing are equivalent through the current `Trace` and the current abstraction works well until we introduce resources.  

- Resource tracing can be implemented either in a stateful subclass without local instances, or in the `Trace` constraint with an awkward signature and weird semantics for the transformers.

- If we resort to a subclass, libraries would be choosing the tradeoff on behalf of apps.

Reply:
I see, thanks, it took me a while to grasp this. IIUC, for a "stateful" instance if you have
```scala
def spanR[A](name: String)(r: Resource[F, A]): Resource[F, A]
```
then you can get "for free"
```scala
def spanR(name: String): Resource[F, Unit] = spanR(name)(Resource.unit)
```
with the exact same semantics as if it had been defined in terms of `spanR(name: String): Resource[F, Unit]` to begin with.

However, for a stateless instance there _is_ a difference whether you use `spanR[A](name: String)(r: Resource[F, A])` or `spanR(name: String): Resource[F, Unit]`. So you want to use the former whenever possible, and only use the latter when necessary e.g. `Stream` and transformer instances.

So in summary, defining `Trace` as in this PR doesn't compromise the power of the stateful `IOLocal` instance, it will work exactly like it does for a Bayou-style `Trace`.

What it does compromise however is the library experience: since we want `spanR(name: String): Resource[F, Unit]` to be the second-choice, we don't directly offer it on the typeclass, even though it's easily derived in practice.

Reply:
Yes:

|                       | acquire | release | use |
|-----------------------|---------|---------|-----|
| Stateful base         | ✅       | ✅       | ✅   |
| Stateful transformed  | ✅       | ✅       | ✅   |
| Stateless base        | 🤷       | 🤷       | 🚫   |
| Stateless transformed | 🚫       | 🚫       | 🚫   |

✅ - both `spanR`
🚫 - neither `spanR`
🤷 - `spanR(name: String)(r: Resource[F, A])`, but not `spanR(name: String)`

Reply:
Coo, thanksl. This PR seems good then: it doesn't compromise the stateful `IOLocal` instance, while still being at least somewhat compatible with `Kleisli`. On the library side, you can trace any `Resource`-like thing either directly or with `trace.spanR(name)(Resource.unit)`. So everyone should be happy, right?

Reply:
I'm still not convinced it's the right abstraction when we have a type class where a column mixes ✅ and 🚫.

I'll fix the other modules, unmark it as a draft, and we'll see if anyone else has any thoughts.

Reply:
> I'm still not convinced it's the right abstraction when we have a type class where a column mixes ✅ and 🚫.

That would leave us with Stateless I I believe? Where you can span a resource-like thing but can't parent any of the acquire/use/release phases.

FWIW I'd like Stateless I even more. Then we can have `spanR(name: String): Resource[F, Unit]`!

Reply:
Under Stateless I, we can never add fields to any phase of a resource span.  It is never ambient and can't be retrieved.

#527 is the most consistent and this has the broadest support.  I think that's the tradeoff we're stuck with.

Reply:
That's assuming that we can't have "inconsistent" implementations of `Trace`, which I think is fine.

Reply:
Correct.  This supports everything with one type class, with three distinct behaviors.  #527 supports one behavior, but drops the instances that can't support it once we upgrade to a resource constraint.

#527 still feels more right to me.  I'll undraft this and maybe we can find a mediator. 😆